### PR TITLE
@orta => Add favorites to Spotlight.

### LINKED ISSUE #782: Keep local search index up-to-date
Part of #777.

Add and remove favourites (artwork, artists, genes) to the on-device index with the Core Spotlight framework. Be sure to set the `uniqueIdentifier` and `relatedUniqueIdentifier` to the URL of the content on Force so that the search result gets properly counted for in Apple’s ranking system. (See [this](https://developer.apple.com/library/prerelease/ios/documentation/General/Conceptual/AppSearch/AppContent.html#//apple_ref/doc/uid/TP40016308-CH7-SW1) doc.)

### PR DESCRIPTION
**Pull request by alloy**
Fixes #782.

This PR supersedes #803.

### CODE REVIEW

File: Artsy/App/ARAppActivityContinuationDelegate.m [hunk 1]
```diff
@@ -10,15 +10,25 @@ + (void)load
 
 - (BOOL)application:(UIApplication *)application willContinueUserActivityWithType:(NSString *)userActivityType;
 {
-    return [userActivityType isEqualToString:NSUserActivityTypeBrowsingWeb] || [userActivityType hasPrefix:@"net.artsy.artsy."];
+    return [userActivityType isEqualToString:NSUserActivityTypeBrowsingWeb]
+            || (&CSSearchableItemActionType != NULL && [userActivityType isEqualToString:CSSearchableItemActionType])
```
Comment (orta) [inline Artsy/App/ARAppActivityContinuationDelegate.m]: any chance you could make this into a named bool? I didn't get what it was for

```
BOOL isActionType = (&CSSearchableItemActionType != NULL && [userActivityType isEqualToString:CSSearchableItemActionType])
```

Comment (orta) [inline Artsy/App/ARAppActivityContinuationDelegate.m]: same for below

File: Artsy/App/ARSpotlight.m [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+static void
+ARSearchAttributesAddThumbnailData(CSSearchableItemAttributeSet *attributeSet,
+                                   NSURL *thumbnailURL,
+                                   ARSearchAttributesCompletionBlock completion)
+{
+    SDWebImageManager *manager = [SDWebImageManager sharedManager];
+    [manager downloadImageWithURL:thumbnailURL
+                          options:0
+                         progress:nil
+                        completed:^(UIImage *image, NSError *_, SDImageCacheType __, BOOL ____, NSURL *_____) {
```
Comment (orta) [inline Artsy/App/ARSpotlight.m]: hah `____`

File: Artsy/App/ARSpotlight.m [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    ARSpotlightAvailable = NSClassFromString(@"CSSearchableIndex") != nil && [CSSearchableIndex isIndexingAvailable];
+
+    if (ARSpotlightAvailable) {
+        ARSearchableIndex = [CSSearchableIndex defaultSearchableIndex];
+
+        ARSpotlightQueue = dispatch_queue_create("net.artsy.artsy.ARSpotlightQueue", DISPATCH_QUEUE_SERIAL);
+
+        // Load/Initialize ARIndexedEntities db.
+        ar_dispatch_on_queue(ARSpotlightQueue, ^{
+            NSString *appSupportDir = [NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory,
```
Comment (orta) [inline Artsy/App/ARSpotlight.m]: optionally you could use `ARFileUtils` for some of this stuff

File: Artsy/App/ARSpotlight.m [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+            CSSearchableItem *item = [[CSSearchableItem alloc] initWithUniqueIdentifier:identifier
+                                                                       domainIdentifier:domainIdentifier
+                                                                           attributeSet:attributeSet];
+            [self.searchableIndex indexSearchableItems:@[item] completionHandler:^(NSError *error) {
+                if (error) {
+                    ARErrorLog(@"Failed to index entity `%@': %@", identifier, error);
+                } else {
+                    ar_dispatch_on_queue(ARSpotlightQueue, ^{
+                        [self.indexedEntities addObject:identifier];
+                        [self.indexedEntities.allObjects writeToFile:ARIndexedEntitiesFile atomically:YES];
```
Comment (orta) [inline Artsy/App/ARSpotlight.m]: this feels like something that could cause performance issues when you have a bunch of favourites

File: Artsy/App/ARSpotlight.m [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+            }
+        }];
+    });
+}
+
+#pragma mark - CSSearchableItemAttributeSet
+
+//
+// Do NOT modify any of the returned CSSearchableItemAttributeSet objects on another queue than ARSpotlightQueue.
+//
```
Comment (orta) [inline Artsy/App/ARSpotlight.m]: :+1:

File: Artsy/App/ARUserActivity.m [hunk 6]
```diff
@@ -1,236 +1,141 @@
...
+NSString *const ARUserActivityTypeShow = @"net.artsy.artsy.show";
 
 
 @implementation ARUserActivity
 
-+ (instancetype)activityWithArtwork:(Artwork *)artwork becomeCurrent:(BOOL)becomeCurrent
+// Do NOT assign a relatedUniqueIdentifier to the attribute set when combining with a user activity.
+// This needs to be done because of: https://forums.developer.apple.com/message/28220#28220
+//
+// TODO Test if this is still an issue in the iOS 9 GM and, if so, add a unit test.
```
Comment (orta) [inline Artsy/App/ARUserActivity.m]: We're at GM now

File: Artsy/App/ARSpotlight.m (hunk 4 above)
Comment (alloy) [inline Artsy/App/ARSpotlight.m]: Possibly. I have not noticed it in my testing, but maybe I don’t have enough favorites. The idea is it being a proper journal, though, so it should be up-to-date if, for instance, the app crashes/quits during the process.

So maybe ask someone with lots of favs to test this during QA? Also while the app is in the background?

File: Artsy/App/ARUserActivity.m (hunk 6 above)
Comment (alloy) [inline Artsy/App/ARUserActivity.m]: Indeed, and it still applies, so I’ll remove that comment.

File: Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+        it(@"includes the artwork title", ^{
+            expect(activity.title).to.equal(@"Artwork Title");
+        });
+
+        it(@"includes additional metadata", ^{
+            expect(activity.userInfo).to.equal(@{@"id": @"artwork_id" });
+        });
+
+        it(@"is eligible for everything", ^{
+            expect(activity.isEligibleForPublicIndexing).to.beTruthy;
```
Comment (orta) [inline Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m]: these need `()` at the end, it's an LLVM bug that you don't get a warning

File: Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+        });
+
+        it(@"includes additional metadata", ^{
+            expect(activity.userInfo).to.equal(@{@"id": @"some-artist" });
+        });
+
+        it(@"is eligible for everything", ^{
+            expect(activity.isEligibleForPublicIndexing).to.beTruthy;
+            expect(activity.isEligibleForSearch).to.beTruthy;
+            expect(activity.isEligibleForHandoff).to.beTruthy;
```
Comment (orta) [inline Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m]: same, best do a quick `find all` for `beTruthy;`

File: Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+            expect(attributeSet.title).to.equal(@"The Fair Affair");
+        });
+
+        it(@"includes the date range of when the fair is running", ^{
+            expect(attributeSet.startDate).to.equal([model startDate]);
+            expect(attributeSet.endDate).to.equal([model endDate]);
+        });
+
+        it(@"does not include a thumbnail without a profile", ^{
+            expect(attributeSet.thumbnailData).to.beNil;
```
Comment (orta) [inline Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m]: here too

File: Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+            attributeSet = [ARSpotlight searchAttributesWithShow:model
+                                                          inFair:nil
+                                               includeIdentifier:NO
+                                                      completion:^(CSSearchableItemAttributeSet *_) {}];
+            expect(activity.contentAttributeSet.description).to.equal(attributeSet.description);
+        });
+    });
+});
+
+describe(@"Indexing favorites", ^{
```
Comment (orta) [inline Artsy_Tests/App_Tests/ARUserActivityAndSpotlightTests.m]: favourites :gb:

File: Artsy/App/ARSpotlight.m (hunk 4 above)
Comment (orta) [inline Artsy/App/ARSpotlight.m]: Yeah, let's get @alexanderfabry to definitely give this a run on his phone, if I recall he has a serious amount of favourites - though I do too but my phone is generally top of the line in performance.

**Comment by orta** [comment]
I felt that everything was readable and made sense. `ARSpotlight.m` was starting to get pretty long, could turn it into two classes, one for the indexing and the other for the Model -> CSIndexableItem stuff. Not a blocker though, from my perspective. 

Waiting on the `()`s for those tests then you're :+1:

File: Artsy/App/ARSpotlight.m (hunk 4 above)
Comment (alloy) [inline Artsy/App/ARSpotlight.m]: :+1:

**Comment by alloy** [comment]
@orta Updated.

**Comment by orta** [comment]
All looks :+1: - merge on the Green and make me a betttaaaaaaaa

**Comment by orta** [comment]
Giphy

**Pull request merged**